numpy/random/mtrand.pyx

```diff
@@ -4468,9 +4468,9 @@ cdef class RandomState:
                     j = random_interval(&self._bitgen, i)
                     if i == j:
                         continue  # i == j is not needed and memcpy is undefined.
-                    buf[...] = x[j]
-                    x[j] = x[i]
-                    x[i] = buf
+                    buf[...] = x[j, ...]
```

Comment:
This is a bug but it might break the stream if the shuffled order may be different.  If it would change the order of the outputs then I think it isn't allowed under the NEP.

" Bug fixes that change the stream SHALL NOT be made to RandomState; instead, buggy distributions should be made to warn when they are buggy. "

Sounds like there should be a warning here rather than a fix.  Although it this would lead to a segfault than it might fall into a grey zone.

Reply:
Hmmm, I had not thought about the stability NEP.  But this will not actually change the order or stream (I guess with a broken array subcalls order changes might be possible, but I doubt it, even matrix is fine in this particular case, but I should add a test for matrix probably!), it will ensure that the input values are actually without mutating (i.e. converting `np.array(4)` into a `4` or so).

Reply:
It comes down to what does it mean that the stream is the same.  While this cannot change the number of random integers needed for the shuffle, the output could change in some cases as you noted (even if it is more correct).  This is a bit ambiguous as to whether this is a different stream.  It probably isn't, but then again, one of the reasons for the NEP was to remove maintenance burden for RandomState.

Reply:
I don't think this change counts as breaking the stream

Reply:
The real controlling statement in NEP 19 is the following: "It is intended that this class will no longer be modified, except to keep it working when numpy internals change." The following sentence about bug fixes is more to counter the natural urge to fix bugs. The narrowness of "bug changes that change the stream" is not meant to limit the changes that are forbidden by the NEP.

The purpose of freezing `RandomState` is to keep its current behavior intact, even the bugs. "RNG stream" was meant to refer to all of the outputs, not just the sequence of `BitGenerator` states.

There is a narrow exception for catastrophic things like segfaults. Does this bug risk a segfault?

Reply:
It could cause a segfault in a caller who has an object array of 0D arrays, and assumes that the elements will still be arrays after shuffling.

Reply:
> The following sentence about bug fixes is more to counter the natural urge to fix bugs.

Or more clearly, it was meant to highlight that the one exception that was allowed under the old policy (bug fixes) is no longer allowed. This change wouldn't have been allowed even under the old policy (if there's no segfault). NEP 19 didn't loosen up the policy for `RandomState`.

Reply:
Does this only show up when the outer array is an `ndarray` subclass (and not `ndarray` itself)? That's the only condition I can get the provided test code to fail on.

The segfault exception is meant more for segfaults happening in the method itself. If it's just that the current documentation of the `RandomState` method is claiming guarantees that cause you to write code around it that might segfault (working at the C API level), then a reasonable fix is to document the current behavior. If pure Python code might invoke the segfault, that leans more into the allowable territory; pure Python code really should never segfault (`ctypes` notwithstanding).

Overall, the behavior change is minor (since it only alters the types and possible copies of the output items), so the risk of segfaults might weigh more heavily. If someone has actually made their C/Cython code segfault with the current behavior, that would tip me into allowing it for `RandomState`. If not, I lean towards documenting it.

Definitely a good fix for `Generator`, though.

Reply:
Well, it is a super niche case of course, it affects 1-d array subclasses and mainly (or possibly only) those containing Python objects.  As is, the current code can _modify_ items in strange ways in addition to shuffling them.

I guess I realized NEP 19 to be quite that strict and considering that many users probably won't switch for years, not sure about the particular tradeoffs... But of course there are also good reasons for being strict.

So just back it out?

Reply:
I think so. Worth documenting in `RandomState`, though.

Reply:
NEP 19 says that if we choose not to fix bugs, we should issue a warning, right?

Reply:
OK, I will look into it. I realized, that I did not mention the weirder parts of it (mainly because I did not realize it was _this_ weird), so posting it for the record:

```
In [1]: arr = np.array(range(100), dtype=object).view(np.ma.MaskedArray)

In [2]: arr
Out[2]: 
masked_array(data=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
                   16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29,
                   30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
                   44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
                   58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,
                   72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85,
                   86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99],
             mask=False,
       fill_value='?',
            dtype=object)

In [3]: np.random.shuffle(arr)

In [4]: arr
Out[4]: 
masked_array(data=[2, 85, array(19, dtype=object),
                   array(19, dtype=object), 49, array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object), 66,
                   58, array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), 87, array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object), array(19, dtype=object),
                   array(19, dtype=object)],
             mask=False,
       fill_value='?',
            dtype=object)
```

Reply:
Yeah, I was thinking of a warning as well now to be honest. It feels too obscure to be useful documented, and to severe to ignore in any case (in the rare case where you _do_ hit it).  It is probably limited to 1-d subclasses containing objects, or structured dtypes ones with a single object field.

Reply:
OK, made a slightly obnoxious warning, but happy to move it or something similar to the docs instead if that feels better.